### ISSUE TITLE
GitHub Action to deploy a release to npm

### ISSUE THREAD

**Issue by Jolg42** [opentypejs/opentype.js#541]
@ILOVEPIE what do you think about a GitHub Actions deployment, I can add the NPM_TOKEN there later for the npm package.

I used this recently on another project, and it could look like this maybe, any other idea is welcome 🙌🏼 

## Publishing a new version on npm

1. Bump the package version
   - `npm --no-git-tag-version version minor` (or `patch` for a patch release)
1. Commit files to main or merge to main via a PR
   - `git add package*`
   - `git commit -am "Release <version>"`
   - `git push`
1. Create a release in the UI -> https://github.com/prisma/opentype.js/releases/new
   - Add a tag, use the version, like `1.0.1`
     - Click on "Create new tag: <version> on publish"
   - Add a release title, use the version, like `1.0.1`
   - Add in the description the changes and links to PRs
   - "Target" should be "main" (default)
   - "Set as the latest release" should be checked (default)
   - Click "Publish release"
   - Monitor https://github.com/prisma/opentype.js/actions/workflows/npm-publish.yml for the publish workflow

And the action
```npm-publish.yml
name: Publish package to npm
on:
  release:
    types: [created]
jobs:
  publish:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3

      # Setup .npmrc file to publish to npm
      - uses: actions/setup-node@v3
        with:
          node-version: '14'
          registry-url: 'https://registry.npmjs.org'

      - run: npm ci

      - run: npm run build

      # Publish to https://www.npmjs.com/package/checkpoint-client
      - run: npm publish
        env:
          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
```

**Comment by ILOVEPIE** [comment]
shouldn't we also have a action that runs dist upon a PR getting merged, also, i activated discussions so that we can keep the issues section dedicated to issues, and discussions separate.

**Comment by ILOVEPIE** [comment]
I don't mean publish on a pr getting merged, just dist, so that the repository contains the lastest build at all times.

**Comment by Jolg42** [comment]
Another option would be to get rid of dist in the Git repository, it would make things easier.
Not sure if anyone is using the dist files, if someone does, we could point them to npm maybe? 
https://www.npmjs.com/package/opentype.js?activeTab=explore (it's a new tab where the code is available)

**Comment by ILOVEPIE** [comment]
The dist files are useful for people using non-npm based systems.

**Comment by ILOVEPIE** [comment]
plus they allow people to test the latest build from the current source code.

**Comment by ILOVEPIE** [comment]
@Jolg42  How do we go about getting this setup?

**Comment by yne** [comment]
I'm voting for the /dist/ folder removal because :

- It makes the merging process impossible since all PR conflict with each other as soon as one od them changed the .min.js version
- People who don't use `npm` can still fetch from the github release section, or from a NPM CDN (unpkg, cloudflare, ...)
- There is not proof that the PR don't contain a malware in the .min.js version
- Git is not meant to store build artefacts

**Comment by ILOVEPIE** [comment]
@yne fair enough, i'll do a PR that adds it to the .gitignore. My question was more on the github action for releasing to NPM & github releases.

**Comment by yne** [comment]
- Github Action shall report status on PR open/update to help reviewer (+ add built artefact to the PR?)
- Github Action shall publish to NPM on [tag creation](https://github.com/opentypejs/opentype.js/releases/new)
- `/dist/` shall be purged (not added to `.gitignore`)

This allow the following use cases:
- If a PR fix a bug: we merge it and add an tag with an incremented micro
- If a PR add a feature: we merge it and an tag with an incremented minor

If this tagging process is correctly done, I don't see why an end-users would need an untagged version.

**Comment by ILOVEPIE** [comment]
@Connum said:
> I think it's good to have the current state of the master available in the dist folder as a preview version for people who want to test it out but don't want to clone and build themselves. This might even help detecting bugs earlier.
> 
> But I also acknowledge that it's annoying to have the files cause trouble when switching branches during development. I guess the optimal way would be to have a github action build the dist files on every merge on the master branch.
> 
> Anyway, if we decide to ignore the dist files, we should also delete the ones that are already tracked before that, otherwise we'll have the outdated files in the repo forever.

I can see both sides on this one. I'm thinking maybe we have a github action that builds a new "release" (labeled something like major.minor.patch-git-commit_hash) every time a pr is merged but doesn't mark it as latest and then when we actually do tag a release it will create a release (labeled major.minor.patch) and mark it as latest. IDK, whatever you think is best.

**Comment by Connum** [comment]
> If this tagging process is correctly done, I don't see why an end-users would need an untagged version.

Due to breaking changes, we're currently on our way to a major version update. It'd be great to have advanced users test it before officially releasing it.

**Comment by ILOVEPIE** [comment]
> Due to breaking changes, we're currently on our way to a major version update. It'd be great to have advanced users test it before officially releasing it.

That's what release candidates are for no?

**Comment by Connum** [comment]
Sure, if those don't accidentally land on npm, that's fine. Though for me an RC is more like "almost ready, we'll only fix bugs and won't add any more feature", but we could also have -dev versions or similar.

**Comment by ILOVEPIE** [comment]
That's what I was suggesting the "releases" ending in -git-commit_hash would be.

**Comment by yne** [comment]
> It'd be great to have advanced users test it before officially releasing it.

@Connum those *advanced* users, who are capable of debugging OpenTypeJS, are probably also capable of running an `npm run build` command.

If they are only supposed to do some testing, then the alpha / beta / RC lifecycle @ILOVEPIE proposed seems very reasonable.

**Comment by Connum** [comment]
Yeah, fair enough! 😄

**Comment by yne** [comment]
I made a POC of https://github.com/yne/opentype.js with
- build + lint + test on PR or merge (fail at `lint` stage ATM)
- NPM release upon git tag
- no more `/dist`

I also took this opportunity to rework the buildflow to use esbuild / eslint / mocha.
- `package.json` now have only 3 devDepdendencies (and 0 vulnerabilities)
- the `docs` folder hold the github page (much cleaner git structure)
- the `src/opentype.js` work out of the box as ESModule (`<script type=module>import ...`), no need to bundle/build it.

Overall I'm really surprised how smooth this migration went (thanks to the clean opentypejs codebase)

@Connum what does the major upgrade brings ? could it be used to migrate this buildflow ?

**Comment by Connum** [comment]
Thanks for the work you put in!
We tagged the issues and PRs we want to ship with the next version with a milestone here: https://github.com/opentypejs/opentype.js/milestone/2

If I see it correctly, you're currently targeting only modern clients with latest ES support, but I guess with esbuild it shouldn't be hard to provide a bundle for older browsers and non-module as well?

**Comment by yne** [comment]
Exactly, esbuild can output both `esm` and legacy (aka `global`) formats.

But shall we move the default entrypoint to the ESM for the 2.0.0 ?

using legacy `--global-name=opentype` format:
```html
<script src="/dist/opentype.global.js"></script>
<script>
fetch('test/fonts/FiraSansMedium.woff')
.then(res => res.arrayBuffer())
.then(buf => console.log(opentype.parse(buf)))
</script>
```

using `--format=esm` we can use the `import` syntax (also directly work on `/src/opentype.js`):
```html
<script type=module>
import * as opentype from '/dist/opentype.esm.js'; // alt: '/src/opentype.js'
fetch('test/fonts/FiraSansMedium.woff')
.then(res => res.arrayBuffer())
.then(buf => console.log(opentype.parse(buf)))
</script>
```

My only concern is that, to share the same ESM build across browsers and node.js, I had to strip usage of `require(fs)` from `load()` and `download()` functions. So the loading look like this for both platform:

```js
// node
const buf = await require('fs').readFile('font.woff');
opentype.parse(buf)

// browser
const buf = await fetch('font.woff').then(res => res.arrayBuffer())
opentype.parse(buf)
```

I can try to make load/download work, but I'm not a big fan of having platform-specific API in the codebase, especially if it only take 1 line from the caller (see: `res.arrayBuffer()` / `fs.readFile`) to do it on it side.

Again, that's a breaking change :/

**Comment by Connum** [comment]
I really wouldn't like seeing that functionality stripped. How do other libraries handle this?

**Comment by yne** [comment]
I can't find much popular hybrid file-processing oriented JS libraries. But here are some:
- [mermaid](https://github.com/mermaid-js/mermaid) 53K :star: : platform-specific are separated in a [cli](https://github.com/mermaid-js/mermaid-cli) repo
- [marked.js](https://github.com/markedjs/marked/blob/master/bin/marked.js) 29K :star: :  `require('fs')` is limited to node-specific `/bin` wrapper script
- [GraphicsMagick](https://github.com/aheckmann/gm#buffers) 6.8K :star: : `require('fs')` is limited to examples and tests. Not in src
- [prettier](https://github.com/prettier/prettier/blob/main/tsconfig.json#L23) 44.8K :star: separate node-related modules at build (I guess ?)

In the meantime I'll continue to try to make `require('fs')` work if ESM version is imported in a browser enviroment.

**Comment by ILOVEPIE** [comment]
@yne I think we should maintain having both an esm and a non-esm build, the current system is expected by a lot of projects, we should also maintain the same naming scheme for the files.
namely ``opentype.js`` and ``opentype.min.js`` being the non-esm version and ``opentype.module.js`` being the esm version.

**Comment by yne** [comment]
@ILOVEPIE you are right. I'll keep the original naming.

**Comment by Connum** [comment]
I'm currently working on SVG-related stuff and noticed today that when we remove the dist/ folder, we'll be no longer able to use unicode-org/text-rendering-tests with pre-release versions by adding them as a github link in the package.json of the tool, but will have to replace the opentype.js file manually inside the node_modules folder.
Absolute edge-case for almost anybody, but I didn't want to let it go completely unnoticed.

**Comment by yne** [comment]
@Connum I dont understand.
I see in they package.json that they pull opentype.js so they will still get the dist files ...

did i missed something from they repo ?

**Comment by Connum** [comment]
It won't be a problem when it's pulled from npm, only when testing unreleased versions locally for generating reports, like I do for https://github.com/Connum/opentypejs-reports

As I said, it's not a big deal, I'll just build the files manually and copy them over.

**Comment by yne** [comment]
You could even automate this build using a [github action](https://github.com/yne/opentype.yne.js/blob/master/.github/workflows/page.yml) that build opentype dist file and deploy the static website (with your built files)

**Comment by Connum** [comment]
That's a good idea!

**Comment by yne** [comment]
https://github.com/opentypejs/opentype.js/pull/567 is ready for review.
- The dist naming are kept, even if `.module.js` shall be .mjs to work in node
- `.parse()` accept both `fs.readFile()` `Buffer` and `fetch()` `ArrayBuffer`  See [loading-a-font](https://github.com/yne/opentype.yne.js/#loading-a-font) example
- The `load`/`loadFromUrl`/`loadFromFile` are kept, even if they are even more useless/absurd now that it can be done in 1 line

Any feedback are welcome

**Comment by Connum** [comment]
Great! I would keep the `.load()` example in the README as an addition, though, as not everyone uses modules yet and can use await in the main level of the file, which might lead to confusion.

**Comment by yne** [comment]
Understood, I'll avoid the await/async syntax.
I'll also add an example to show how to load from an `<input type=file>`

**Comment by yne** [comment]
Okay, I documented both `.parse()` / `.load()` methods (with a bit of warning on the `.load()`)
and both style (`async` / `sync`)

https://github.com/yne/opentype.yne.js#loading-a-font

**Comment by yne** [comment]
@Jolg42 You need to set the `NPM_TOKEN` secret to enable tag publishing.

Once set we could close this issue

**Comment by Connum** [comment]
Pinging @Jolg42 - would be great to be able to make new releases

**Comment by Jolg42** [comment]
@Connum What do you think about this?

I modified the publish script and made it more flexible, see https://github.com/opentypejs/opentype.js/commit/3cfe131c7116edaa554873d788e2b414906d572c

Here is the first run, (a dry run): 
https://github.com/opentypejs/opentype.js/actions/runs/6761331868

All runs will be listed at https://github.com/opentypejs/opentype.js/actions/workflows/release.yml

I already added the NPM_TOKEN, so to publish to npm, leave the dry run checkbox unchecked and that would do it 🤞🏼.

**Comment by Connum** [comment]
Great to be able to do a dry run! But I don't quite get the first version field. The second one is passed as the tag parameter to `npm publish`, that's clear. But why does that first one have to be specified and will be set via `npm pkg set` during the workflow (which will then not be reflected in the repo)? Shouldn't we set the correct version in the repos package.json before publishing anything, anyway?

**Comment by yne** [comment]
As I understand, the "version" value in git's package.json is moot.
The real value is set at publish time by the CI.
This way the CI is in charge of the tagging (not the package.json)

A PR can't set the version, nor multiple commit/GHaction can publish the same package.json version

Kinda agree with this workflow.

**Comment by Connum** [comment]
@Jolg42 could you clarify the two version fields? I think than this would be ready to go!

**Comment by ILOVEPIE** [comment]
@Jolg42 @yne @Connum  We should probably get this figgured out before the release of 2.0.0. And we want to get 2.0.0 out as fast as possible.
